lib/Dialect/FIRRTL/Transforms/CheckRecursiveInstantiation.cpp

```diff
@@ -0,0 +1,91 @@
+//===- CheckRecursiveInstantiation.cpp - Check recurisve instantiation ----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/FIRRTL/FIRRTLInstanceGraph.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/SCCIterator.h"
+
+namespace circt {
+namespace firrtl {
+#define GEN_PASS_DEF_CHECKRECURSIVEINSTANTIATION
+#include "circt/Dialect/FIRRTL/Passes.h.inc"
+} // namespace firrtl
+} // namespace circt
+
+using namespace circt;
+using namespace firrtl;
+
+static void printPath(InstanceGraph &instanceGraph,
+                      ArrayRef<InstanceGraphNode *> nodes) {
+  assert(nodes.size() > 0 && "an scc should have at least one node");
+
+  // The first node in the SCC where we start the path from.
+  auto *first = nodes.front();
+  auto diag = emitError(first->getModule().getLoc(), "recursive instantiation");
+
+  // A set of all nodes in the SCC for quick lookup.
+  llvm::SmallPtrSet<InstanceGraphNode *, 8> scc(nodes.begin(), nodes.end());
+
+  // A set of all nodes seen in the path.
+  llvm::SmallPtrSet<InstanceGraphNode *, 8> visited;
+
+  // The current node we are examining.
+  auto *current = first;
+  while (true) {
+    // Iterate all instances in this module.
+    for (auto *record : *current) {
+      auto *target = record->getTarget();
+      // Filter out the edges that don't point in the SCC.
+      if (!scc.contains(target))
+        continue;
+      // If this node is already on our path, skip it.
+      if (visited.contains(target))
+        continue;
+      auto &note = diag.attachNote(record->getInstance().getLoc());
+      note << record->getParent()->getModule().getModuleName();
+      note << " instantiates "
+           << record->getTarget()->getModule().getModuleName() << " here";
+      // If we have looped back to the starting point, return here.
+      if (target == first)
```

Comment:
I'm not sure about this?

The following causes an infinite loop:
```
firrtl.circuit "SCCMore" {
  firrtl.module @SCCMore() {
    firrtl.instance a @A()
  }
  firrtl.module @A() {
    firrtl.instance b @B()
    firrtl.instance c @C()
  }
  firrtl.module @B() {
    firrtl.instance c @C()
    firrtl.instance d @D()
  }

  firrtl.module @C() {
    firrtl.instance a @A()
    firrtl.instance b @B()
  }

  firrtl.module @D() {
    firrtl.instance a @A()
    firrtl.instance c @C()
  }
}
```

(may not be minimal)

Reply:
I had hastily adapted some code from clang-tidy, while trying to make everything simpler (replacing my proper depth first path search), which didn't work quite the way I expected.  I have decided I don't like how clang-tidy reports these errors, as the function they attach the error to does not necessarily show up in the recursive call chain they report.  So I ripped out this code and went with something even simpler, we just report all modules and instances which are involved in the SCC. I added a smaller version of your test case.